So now our linter is forcing bugs on us? 💢 
```
./lib/galaxy/datatypes/tabular.py:1280:17: E741 ambiguous variable name 'l'
./lib/galaxy/datatypes/tabular.py:1284:25: E741 ambiguous variable name 'l'
```

--- a/lib/galaxy/datatypes/tabular.py
+++ b/lib/galaxy/datatypes/tabular.py
@@ -1277,6 +1277,7 @@ class MatrixMarket(TabularData):
         if dataset.has_data():
             # If the dataset is larger than optional_metadata, just count comment lines.
             with open(dataset.file_name) as dataset_fh:
+                l = ''
                 comment_lines = 0
                 if self.max_optional_metadata_filesize >= 0 and dataset.get_size() > self.max_optional_metadata_filesize:
                     # If the dataset is larger than optional_metadata, just count comment lines.

--- a/lib/galaxy/datatypes/tabular.py
+++ b/lib/galaxy/datatypes/tabular.py
@@ -1281,8 +1281,8 @@ class MatrixMarket(TabularData):
                 comment_lines = 0
                 if self.max_optional_metadata_filesize >= 0 and dataset.get_size() > self.max_optional_metadata_filesize:
                     # If the dataset is larger than optional_metadata, just count comment lines.
-                    for line in dataset_fh:
-                        if line.startswith('%'):
+                    for l in dataset_fh:
+                        if l.startswith('%'):
                             comment_lines += 1
                         else:
                             # No more comments, and the file is too big to look at the whole thing. Give up.

--- a/lib/galaxy/datatypes/tabular.py
+++ b/lib/galaxy/datatypes/tabular.py
@@ -1277,26 +1277,23 @@ class MatrixMarket(TabularData):
         if dataset.has_data():
             # If the dataset is larger than optional_metadata, just count comment lines.
             with open(dataset.file_name) as dataset_fh:
-                l = ''
+                line = ''
+                data_lines = 0
                 comment_lines = 0
-                if self.max_optional_metadata_filesize >= 0 and dataset.get_size() > self.max_optional_metadata_filesize:
-                    # If the dataset is larger than optional_metadata, just count comment lines.
-                    for l in dataset_fh:
-                        if l.startswith('%'):
-                            comment_lines += 1
-                        else:
-                            # No more comments, and the file is too big to look at the whole thing. Give up.
-                            dataset.metadata.data_lines = None
-                            break
-                else:
-                    for i, l in enumerate(dataset_fh):
-                        if l.startswith('%'):
-                            comment_lines += 1
-                    dataset.metadata.data_lines = i + 1 - comment_lines
-                if ' ' in l:
+                # If the dataset is larger than optional_metadata, just count comment lines.
+                count_comments_only = self.max_optional_metadata_filesize >= 0 and dataset.get_size() > self.max_optional_metadata_filesize
+                for line in dataset_fh:
+                    if line.startswith('%'):
+                        comment_lines += 1
+                    elif count_comments_only:
+                        break
+                    else:
+                        data_lines += 1
+                if ' ' in line:
                     dataset.metadata.delimiter = ' '
                 else:
                     dataset.metadata.delimiter = '\t'
             dataset.metadata.comment_lines = comment_lines
+            dataset.metadata.data_lines = data_lines + comment_lines if data_lines else None
             dataset.metadata.columns = 3
             dataset.metadata.column_types = ['int', 'int', 'float']

